File: merge_sort/scripts/merge_sort.js
const fs = require('fs');
const { performance } = require('perf_hooks');

function mergeSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }
    const mid = Math.floor(arr.length / 2);
    const left = mergeSort(arr.slice(0, mid));
    const right = mergeSort(arr.slice(mid));
    return merge(left, right);
}

function merge(left, right) {
    let result = [];
    let leftIndex = 0;
    let rightIndex = 0;
    while (leftIndex < left.length && rightIndex < right.length) {
        if (left[leftIndex] < right[rightIndex]) {
            result.push(left[leftIndex]);
            leftIndex++;
        } else {
            result.push(right[rightIndex]);
            rightIndex++;
        }
    }
    return result.concat(left.slice(leftIndex)).concat(right.slice(rightIndex));
}

fs.readFile('/data/shared_list.txt', 'utf8', (err, data) => {
    if (err) throw err;
    const arr = data.trim().split(',').map(Number);

    const startTime = performance.now();
    const startUsage = process.cpuUsage();
    const startMemory = process.memoryUsage().heapUsed;

    mergeSort(arr);

    const endTime = performance.now();
    const endUsage = process.cpuUsage(startUsage);
    const endMemory = process.memoryUsage().heapUsed;

    fs.writeFileSync('/data/results_js.txt', `Execution time: ${(endTime - startTime) / 1000} seconds\n`);
    fs.writeFileSync('/data/results_js.txt', `CPU time: ${(endUsage.user + endUsage.system) / 1000000} seconds\n`, { flag: 'a' });
    fs.writeFileSync('/data/results_js.txt', `Memory usage: ${endMemory - startMemory} bytes\n`, { flag: 'a' });
});

File: merge_sort/scripts/merge_sort.py
import time
import os
import psutil

def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]
        merge_sort(L)
        merge_sort(R)
        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1
        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

if __name__ == "__main__":
    with open('/data/shared_list.txt', 'r') as f:
        arr = list(map(int, f.readline().strip().split(',')))

    start_time = time.time()
    process = psutil.Process(os.getpid())
    cpu_times_start = process.cpu_times()
    memory_start = process.memory_info().rss

    merge_sort(arr)

    end_time = time.time()
    cpu_times_end = process.cpu_times()
    memory_end = process.memory_info().rss

    with open('/data/results_python.txt', 'w') as f:
        f.write(f"Execution time: {end_time - start_time} seconds\n")
        f.write(f"CPU time: {cpu_times_end.user - cpu_times_start.user} seconds\n")
        f.write(f"Memory usage: {memory_end - memory_start} bytes\n")
